### PR TITLE
KAFKA-13958: Expose logdirs total and usable space via Kafka API (KIP…

### PR DESCRIPTION
**Pull request by mimaison**
…-827)

This implements KIP-827: https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+logdirs+total+and+usable+space+via+Kafka+API

Add TotalBytes and UsableBytes to DescribeLogDirsResponse
Add matching getters on LogDirDescription

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 1]
```diff
@@ -812,13 +817,16 @@ class ReplicaManager(val config: KafkaConfig,
...
+              .setTotalBytes(totalBytes).setUsableBytes(usableBytes)
           case None =>
             new DescribeLogDirsResponseData.DescribeLogDirsResult().setLogDir(absolutePath)
               .setErrorCode(Errors.NONE.code)
+              .setTotalBytes(totalBytes).setUsableBytes(usableBytes)
         }
 
       } catch {
         case e: KafkaStorageException =>
+          e.printStackTrace()
```
Comment (soarez) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: Is this change intended? The exception is already logged in the following line .

**Review by soarez — COMMENTED** [review]
Thanks for this PR

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 1 above)
Comment (mimaison) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: Oops, no we obviously don't want that! Thanks, fixed

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 2]
```diff
@@ -4144,6 +4148,31 @@ class ReplicaManagerTest {
...
+    // Prepare the mocked components for the test
+    val (replicaManager, mockLogMgr) = prepareReplicaManagerAndLogManager(new MockTimer(time),
+      topicPartition, leaderEpoch + leaderEpochIncrement, followerBrokerId, leaderBrokerId, countDownLatch,
+      expectTruncation = false, localLogOffset = Some(10), offsetFromLeader = offsetFromLeader, topicId = Some(topicId))
+
+    val responses = replicaManager.describeLogDirs(Set(new TopicPartition(topic, topicPartition)))
+    assertEquals(mockLogMgr.liveLogDirs.size, responses.size)
+    responses.foreach { response =>
+      assertEquals(Errors.NONE.code, response.errorCode)
+      assertTrue(response.totalBytes >= 0)
```
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Is there a reason due to which totalBytes could be 0? Perhaps we want to  assert a > 0 condition here?

File: clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsResponse.java [hunk 3]
```diff
@@ -31,6 +31,7 @@
 public class DescribeLogDirsResponse extends AbstractResponse {
 
     public static final long INVALID_OFFSET_LAG = -1L;
+    public static final long UNKNOWN_VOLUME_BYTES = -1L;
```
Comment (divijvaidya) [inline clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsResponse.java:34]: Do we still want this? 
Correct me if I am wrong here but I thought that we reached a conclusion in the KIP that Optional will cover the scenario when client is using a newer API and broker is old.

File: clients/src/main/java/org/apache/kafka/clients/admin/LogDirDescription.java [hunk 4]
```diff
@@ -20,19 +20,29 @@
...
  * A description of a log directory on a particular broker.
  */
 public class LogDirDescription {
     private final Map<TopicPartition, ReplicaInfo> replicaInfos;
     private final ApiException error;
+    private final OptionalLong totalBytes;
+    private final OptionalLong usableBytes;
 
     public LogDirDescription(ApiException error, Map<TopicPartition, ReplicaInfo> replicaInfos) {
+        this(error, replicaInfos, UNKNOWN_VOLUME_BYTES, UNKNOWN_VOLUME_BYTES);
```
Comment (divijvaidya) [inline clients/src/main/java/org/apache/kafka/clients/admin/LogDirDescription.java:38]: Instead of UNKNOWN_VOLUME_BYTES, this could be `null` or Optional.Empty()? That way we could get rid of `UNKNOWN_VOLUME_BYTES` completely.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 5]
```diff
@@ -2716,7 +2716,7 @@ private static Map<String, LogDirDescription> logDirDescriptions(DescribeLogDirs
                             new ReplicaInfo(p.partitionSize(), p.offsetLag(), p.isFutureKey()));
                 }
             }
-            result.put(logDirResult.logDir(), new LogDirDescription(Errors.forCode(logDirResult.errorCode()).exception(), replicaInfoMap));
+            result.put(logDirResult.logDir(), new LogDirDescription(Errors.forCode(logDirResult.errorCode()).exception(), replicaInfoMap, logDirResult.totalBytes(), logDirResult.usableBytes()));
```
Comment (tombentley) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: long line

File: clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsResponse.java (hunk 3 above)
Comment (mimaison) [reply clients/src/main/java/org/apache/kafka/common/requests/DescribeLogDirsResponse.java:34]: The wire protocol does not have nullable integers. So the new fields are defined to use -1 as the default value in the protocol: https://github.com/apache/kafka/blob/a73b96a7f6c467b32ca30b93534468fd192eeabe/clients/src/main/resources/common/message/DescribeLogDirsResponse.json#L52-L56

This constant is used to map the value in DescribeLogDirsResponse to an Optional: https://github.com/apache/kafka/blob/a73b96a7f6c467b32ca30b93534468fd192eeabe/clients/src/main/java/org/apache/kafka/clients/admin/LogDirDescription.java#L44-L45

File: clients/src/main/java/org/apache/kafka/clients/admin/LogDirDescription.java [hunk 6]
```diff
@@ -54,11 +64,21 @@ public Map<TopicPartition, ReplicaInfo> replicaInfos() {
         return unmodifiableMap(replicaInfos);
     }
 
+    public OptionalLong totalBytes() {
```
Comment (tombentley) [inline clients/src/main/java/org/apache/kafka/clients/admin/LogDirDescription.java:71]: Javadoc

File: clients/src/main/java/org/apache/kafka/clients/admin/LogDirDescription.java [hunk 7]
```diff
@@ -54,11 +64,21 @@ public Map<TopicPartition, ReplicaInfo> replicaInfos() {
...
+        return totalBytes;
+    }
+
+    public OptionalLong usableBytes() {
```
Comment (tombentley) [inline clients/src/main/java/org/apache/kafka/clients/admin/LogDirDescription.java:79]: Javadoc

Comment (tombentley) [reply clients/src/main/java/org/apache/kafka/clients/admin/LogDirDescription.java:79]: Also do we want to say something about the contraints on usableBytes and totalBytes (individually and compared with each other).

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala (hunk 2 above)
Comment (mimaison) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: True this assertion is a bit too defensive! > 0 should be fine. I'll push an update.

**Review by tombentley — COMMENTED** [review]
Thanks @mimaison

File: clients/src/main/java/org/apache/kafka/clients/admin/LogDirDescription.java (hunk 6 above)
Comment (mimaison) [reply clients/src/main/java/org/apache/kafka/clients/admin/LogDirDescription.java:71]: Good point, added!

File: clients/src/main/java/org/apache/kafka/clients/admin/LogDirDescription.java (hunk 7 above)
Comment (mimaison) [reply clients/src/main/java/org/apache/kafka/clients/admin/LogDirDescription.java:79]: I've added some details, let me know if it's not what you had in mind

**Comment by divijvaidya** [comment]
Thank you for making the suggested changes. 

The current set of changes don't contain a fully end to end integration test (KafkaAdminClientTest uses response mocks). Please make a change in `DescribeLogDirsRequestTest.scala` to validate the new changes.

**Comment by mimaison** [comment]
Thanks @divijvaidya for the review. I've updated the end to end tests to cover this new feature.

**Review by divijvaidya — APPROVED** [review]
LGTM!

**Review by viktorsomogyi — APPROVED** [review]
@mimaison thanks for this feature, it's a really useful addition for us as well.
Tried out your code locally, ran the tests, they all work. I don't have any particular comments, it looks good to me.

**Review by tombentley — APPROVED** [review]
LGTM, thanks @mimaison!
